safehttp/mux.go

```diff
@@ -0,0 +1,156 @@
+	MethodPost = "POST"
+)
+
+// ServeMux TODO
+type ServeMux struct {
+	mux        *http.ServeMux
+	domains    map[string]bool
+	dispatcher Dispatcher
+
+	// Maps user-provided patterns to combined handlers which encapsulate
+	// multiple handlers, each one associated with an HTTP method.
+	handlers map[string]combinedHandler
+}
+
+// NewServeMux returns a safe multiplex containing the list of allowed domains
+// and dispatcher provided by the user.
+func NewServeMux(dispatcher Dispatcher, domains ...string) *ServeMux {
+	d := map[string]bool{}
+	for _, host := range domains {
+		d[host] = true
+	}
+	return &ServeMux{mux: http.NewServeMux(), domains: d, dispatcher: dispatcher, handlers: map[string]combinedHandler{}}
+}
+
+// Handle registers a handler for the given pattern and method. The options
+// are applied to the middleware. If another handler is already registered
+// for the same pattern and method, Handle panics.
+func (m *ServeMux) Handle(pattern string, h map[string]Handler) {
+	ch := combinedHandler{h: h}
+	m.handlers[pattern] = ch
+	m.mux.Handle(pattern, ch.combine(m.domains, m.dispatcher))
+}
+
+// HandleFunc registers a handler for the given pattern and method. The options
+// are applied to the middleware. If another handler is already registered
+// for the same pattern and method, HandleFunc panics.
+func (m *ServeMux) HandleFunc(pattern string, h map[string]HandleFunc) {
+	if h == nil {
+		return
+	}
+	newH := make(map[string]Handler)
+	for key, val := range h {
+		newH[key] = HandlerFunc(val)
+	}
+	m.Handle(pattern, newH)
+}
+
+// NotFound registers a handler which will be called when an unhandled
+// path is visited. If another handler has already been registered for
+// this purpose, NotFound panics.
+func (*ServeMux) NotFound(h Handler) {
+	panic("not implemented")
+}
+
+// NotFoundFunc registers a handler which will be called when an unhandled
+// path is visited. If another handler has already been registered for
+// this purpose, NotFound panics.
+func (*ServeMux) NotFoundFunc(h HandleFunc) {
+	panic("not implemented")
+}
+
+// Handler returns the handler to use for the incoming request and the pattern.
+func (m *ServeMux) Handler(r *IncomingRequest) (Handler, string) {
+	h, pattern := m.mux.Handler(r.req)
+
+	if pattern == "" {
+		// If the pattern is empty, no handler was registered and the handler is
+		// an http.NotFoundHandler
+		// TODO: replace http.NotFoundHandler to a safehttp.NotFoundHandler
+		return safeHandler(h), pattern
+	}
+
+	// See if it is redirect or combined
+	c, ok := m.handlers[pattern]
+	if !ok {
+		// We got a http.RedirectHandler from m.mux.Handler()
+		return safeHandler(h), pattern
+	}
+
+	safeH, err := c.lookup(r.req.Method)
+	if err != nil {
+		// err is nil unless no HTTP method was registered for this pattern
+		// TODO: replace http.NotFoundHandler to a safehttp.MethodNotAllowedHandler
+		return safeHandler(http.NotFoundHandler()), pattern
+	}
+
+	return safeH, pattern
+}
```

Comment:
Can we drop this for now? We don't really need it as it is, and I don't really get the purpose of `safeHandler` yet, and this is the only usage.